src/python/pants/backend/python/goals/setup_py.py

```diff
@@ -589,15 +586,47 @@ async def generate_chroot(request: SetupPyChrootRequest) -> SetupPyChroot:
                 f"`entry_point='{entry_point.module}:main'. See {url}."
             )
         entry_point_requests.append(ResolvePexEntryPointRequest(binary[PexEntryPointField]))
+
     binary_entry_points = await MultiGet(
         Get(ResolvedPexEntryPoint, ResolvePexEntryPointRequest, request)
         for request in entry_point_requests
     )
+    source_with_binaries = dict()
     for key, binary_entry_point in zip(key_to_binary_spec.keys(), binary_entry_points):
-        entry_points = setup_kwargs.setdefault("entry_points", {})
-        console_scripts = entry_points.setdefault("console_scripts", [])
         if binary_entry_point.val is not None:
-            console_scripts.append(f"{key}={binary_entry_point.val.spec}")
+            source_with_binaries[key] = binary_entry_point.val.spec
+
+    # Collect entry points from `python_distribution(entry_points=...)`
+    source_entry_points = dict()
+    if exported_target.target.has_field(PythonDistributionEntryPoints):
+        entry_points_field = await Get(
+            ResolvedPythonDistributionEntryPoints,
+            ResolvePythonDistributionEntryPointsRequest(
+                exported_target.target[PythonDistributionEntryPoints]
+            ),
+        )
+
+        if entry_points_field.val:
+            for key, section in entry_points_field.val.items():
+                source_entry_points[key] = {
+                    name: entry_point.spec for name, entry_point in section.items()
+                }
+
+    # Collect any entry points from the setup_py() object.
+    source_provides = dict()
+    if "entry_points" in setup_kwargs:
+        source_provides = setup_kwargs["entry_points"]
+
+    # Gather entry points with source description for any error messages when merging them.
+    entry_point_sources = {
+        f"{exported_addr} `provides.with_binaries()`": {"console_scripts": source_with_binaries},
+        f"{exported_addr} `entry_points`": source_entry_points,
+        f"{exported_addr} `provides.entry_points`": source_provides,
+    }
```

Comment:
Nice!

Reply:
I'm trying to work out in my head in which order to have these, though. As I guess it will affect the potential error message in case you get merge conflicts..